synapse/storage/databases/main/event_push_actions.py

```diff
@@ -463,6 +464,141 @@ def add_thread_id_summary_txn(txn: LoggingTransaction) -> int:
 
         return result
 
+    async def get_unread_counts_by_room_for_user(self, user_id: str) -> Dict[str, int]:
+        """Get the notification count by room for a user. Only considers notifications,
+        no highlights or unreads, and threads are currently aggregated under their room.
+
+        This function is intentionally not cached because it is called to calculate the
+        unread badge for notifications and thus the result is expected to change.
+
+        Note that this function assumes the user is a member of the room. Because
+        summary rows are not removed when a user leaves a room, the caller must
+        filter out those results from the result.
+        """
+        return await self.db_pool.runInteraction(
+            "get_unread_counts_by_room_for_user",
+            self._get_unread_counts_by_room_for_user_txn,
+            user_id,
+        )
+
+    def _get_unread_counts_by_room_for_user_txn(
+        self, txn: LoggingTransaction, user_id: str
+    ) -> Dict[str, int]:
+        receipt_types_clause, args = make_in_list_sql_clause(
+            self.database_engine,
+            "receipt_type",
+            (ReceiptTypes.READ, ReceiptTypes.READ_PRIVATE),
+        )
+        args.extend([user_id, user_id])
+
+        room_to_count: Dict[str, int] = defaultdict(int)
+
+        # First get summary counts by room / thread for the user. Note we use a OR join
+        # condition here such that we handle both receipts with thread ID's and those
+        # without that get applied to any thread_id values.
+        sql = f"""
+            SELECT eps.room_id, notif_count
+            FROM event_push_summary AS eps
+            LEFT JOIN (
+                SELECT room_id, thread_id, MAX(stream_ordering) AS receipt_stream_ordering
+                FROM receipts_linearized
+                LEFT JOIN events USING (room_id, event_id)
+                WHERE
+                    {receipt_types_clause}
+                    AND user_id = ?
+                GROUP BY room_id, thread_id
+            ) AS receipts ON (
+                eps.room_id = receipts.room_id
+                AND eps.thread_id = receipts.thread_id
+            ) OR (
+                eps.room_id = receipts.room_id
+                AND receipts.thread_id IS NULL
+            )
+            WHERE user_id = ?
+            AND notif_count != 0
+            AND (
+                (last_receipt_stream_ordering IS NULL AND stream_ordering > receipt_stream_ordering)
+                OR last_receipt_stream_ordering = receipt_stream_ordering
+                OR receipt_stream_ordering IS NULL
+            )
```

Comment:
This looks plausible, but I'm trying to convince myself it is correct:

1. There's an inner query which fetches the maximum receipt stream ordering per room / thread. (This will also potentially return an unthreaded receipt stream ordering.)
2. We left join to `event_push_summary` by room and either a matching thread ID or as an unthreaded receipt. This will result in a row per room / thread, with duplicate rows if there's an unthreaded receipt.
3. Filter by user / ensuring there's notifications. ✅ 
4. Attempt to ensure that the summary is up-to-date with the latest receipt.

I'm trying to think through it, but I *think* there's a situation where we would double count the summary due to both the threaded and unthreaded receipt?

Reply:
(I think this is *somewhere* missing a comparison between the unthreaded and threaded read receipt stream ordering, only picking the `max` per thread.)

I had something similar which I backed out in https://github.com/matrix-org/synapse/pull/13877/commits/041fe7f39bca11bcb250e5205ec57822d96bef40 because it was not needed when considering only a single room at once.

You might be able to do something like:

```sql
SELECT eps.room_id, notif_count
FROM event_push_summary AS eps
LEFT JOIN (
    SELECT room_id, thread_id, stream_ordering AS threaded_receipt_stream_ordering
    FROM receipts_linearized
    LEFT JOIN events USING (room_id, event_id)
    WHERE
        {receipt_types_clause}
        AND user_id = ?
        AND thread_id IS NOT NULL
    GROUP BY room_id, thread_id
) AS threaded_receipts USING (room_id, thread_id)
LEFT JOIN (
    SELECT room_id, stream_ordering AS unthreaded_receipt_stream_ordering
    FROM receipts_linearized
    LEFT JOIN events USING (room_id, event_id)
    WHERE
        {receipt_types_clause}
        AND user_id = ?
        AND thread_id IS NULL
    GROUP BY room_id, thread_id
) AS threaded_receipts USING (room_id, thread_id)
WHERE user_id = ?
AND notif_count != 0
AND (
    (last_receipt_stream_ordering IS NULL AND stream_ordering > GREATEST(threaded_receipt_stream_ordering, unthreaded_receipt_stream_ordering))
    OR last_receipt_stream_ordering = GREATEST(threaded_receipt_stream_ordering, unthreaded_receipt_stream_ordering)
)
```

(Note that query is Postgres only, you need to use `MAX` on sqlite.)

Also, I think this needs to properly handle there not being any receipt and knowing when a user joined the room?

Reply:
Struggling to wrap my head around these queries a bit! Started by adding the test to the threaded/mixed counts as well as the old ones: https://github.com/matrix-org/synapse/pull/14255/commits/bd7c69470e6bc4847641ce7932687722f16ba9ec, and then based off your query + results rewrote the queries in https://github.com/matrix-org/synapse/pull/14255/commits/ffaba6058057260d572ef4a0d7ef7d062e87c585 which now passes all the count tests :)